src/c/funkce.md

```diff
@@ -1,148 +1,204 @@
+funkce `vypis_text`. Jakmile se příkazy z této funkce vykonají, tak program bude pokračovat ve funkci
+`main`.
 
-V jazyce *C* máme jeden datový typ, který nemá přesné určení. Je to typ a
-může prakticky znamenat cokoli. Nejčastěji jej používáme jako pointer na
-nějakou struktury, kterou pak budeme dále přetypovávat. Velmi často je
-však používán jako návratový typ funkce. V takovém případě funkce
-nevrací žádnou hodnotu. Funkce pouze zpracuje vstupní data a tím její
-úloha končí. Takové funkci říkáme procedura.
+Pomocí volání funkcí můžeme mít kus kódu v programu zapsán pouze jednou ve funkci, a poté ho
+můžeme spouštět z různých částí programu, podle toho, kdy se nám to zrovna bude hodit.
 
-Parametry funkcí
-----------------
+### Parametrizace funkcí
+Funkcím lze dávat vstupy zvané **parametry** (*parameters*). Parametry jsou proměnné uvnitř funkce,
+jejichž hodnotu nastavujeme při zavolání dané funkce. Například následující funkce `vypis_cislo` má
+parametr `cislo` s datovým typem `int`.
+```c,editable
+#include <stdio.h>
+
+void vypis_cislo(int cislo) {
+    printf("Cislo: %d\n", cislo);
+}
+int main() {
+    vypis_cislo(5);
+    return 0;
+}
+```
+Při zavolání funkce musíme pro každý její parametr do závorek dát hodnotu odpovídajícího datového typu.
+Zde je jediný parameter typu `int`, takže při zavolání této funkce musíme do závorek dát jednu hodnotu
+datového typu `int`: `vypis_cislo(5)`. Před spuštěním příkazů ve funkci dojde k tomu, že hodnota každého
+parametru se nastaví na hodnotu předanou ve volání funkce[^3]. Při zavolání `vypis_cislo(5)` si tak můžete
+představit, že se vykoná následující kód:
+```c
+{
+    // nastavení hodnot parametrů
+    int cislo = 5;
 
-V prvním příkladu pro výpočet druhé mocniny jsme použili parametr tak,
-že byl volán tzv. hodnotou. To prakticky znamená, že hodnota uložená v
-proměnné `number` na řádku 5 se nakopíruje do parametru `number` funkce `sqr` na řádku 1. Snadno si
-lze představit, že pokud bychom chtěli ve funkci pracovat s větším
-objemem dat, vyžadovalo by to velké kopírování dat, což by náš program
-zpomalovalo. Další problém pak může nastat, když by kopírovaná data byla
-větší než velikost stacku, kterou máme přidělenu.
+    // tělo funkce
+    printf("Cislo: %d\n", cislo); 
+}
+```
 
-Vyřešit můžeme tyto problémy jednoduše tak, že budeme argument funkce
-předávat odkazem. Odkaz bude v kontextu našeho jazyka jednoduše pointer.
-Ukažme si, jak upravit funkci pro výpočet druhé mocniny, aby akceptovala
-argument předávaný odkazem.
+[^3]: Hodnoty (výrazy) předávané při volání funkce se nazývají **argumenty** (*arguments*). Při
+volání `vypis_cislo(5)` se tedy do parametru `cislo` nastaví hodnota argumentu `5`.
 
-TODO: Co tohle je za kravina... Predelat!!!
+Parametrů mohou funkce brát libovolný počet, nicméně obvykle se používá maximálně cca 5
+parametrů, aby funkce a její používání (volání) nebylo příliš složité. Jednotlivé parametry jsou
+odděleny v definici funkce i v jejím volání čárkami:
```

Comment:
jeste nejak zminit, ze "formalni nazev promenne argumentu" nesouvisi s pojmenovanim promenne u volajiciho?
to taky byva casto problem

Reply:
Chtěl jsem to zmínit, ale když to rovnou ukazuju na výrazech typu `5`, tak by to snad mělo být jasné, žádné proměnné jsem tam ani nepoužíval. Dal bych to možná spíše do sekce "časté chyby".